apps/condo/migrations/20220704172756-0143_ticket_completedat_ticketchange_completedatfrom_and_more.js

```diff
@@ -0,0 +1,65 @@
+// auto generated by kmigrator
+// KMIGRATOR:0143_ticket_completedat_ticketchange_completedatfrom_and_more:IyBHZW5lcmF0ZWQgYnkgRGphbmdvIDQuMCBvbiAyMDIyLTA3LTA0IDEyOjI3Cgpmcm9tIGRqYW5nby5kYiBpbXBvcnQgbWlncmF0aW9ucywgbW9kZWxzCgoKY2xhc3MgTWlncmF0aW9uKG1pZ3JhdGlvbnMuTWlncmF0aW9uKToKCiAgICBkZXBlbmRlbmNpZXMgPSBbCiAgICAgICAgKCdfZGphbmdvX3NjaGVtYScsICcwMTQyX3BheW1lbnRfb3JkZXJfcGF5bWVudGhpc3RvcnlyZWNvcmRfb3JkZXInKSwKICAgIF0KCiAgICBvcGVyYXRpb25zID0gWwogICAgICAgIG1pZ3JhdGlvbnMuQWRkRmllbGQoCiAgICAgICAgICAgIG1vZGVsX25hbWU9J3RpY2tldCcsCiAgICAgICAgICAgIG5hbWU9J2NvbXBsZXRlZEF0JywKICAgICAgICAgICAgZmllbGQ9bW9kZWxzLkRhdGVUaW1lRmllbGQoYmxhbms9VHJ1ZSwgbnVsbD1UcnVlKSwKICAgICAgICApLAogICAgICAgIG1pZ3JhdGlvbnMuQWRkRmllbGQoCiAgICAgICAgICAgIG1vZGVsX25hbWU9J3RpY2tldGNoYW5nZScsCiAgICAgICAgICAgIG5hbWU9J2NvbXBsZXRlZEF0RnJvbScsCiAgICAgICAgICAgIGZpZWxkPW1vZGVscy5EYXRlVGltZUZpZWxkKGJsYW5rPVRydWUsIG51bGw9VHJ1ZSksCiAgICAgICAgKSwKICAgICAgICBtaWdyYXRpb25zLkFkZEZpZWxkKAogICAgICAgICAgICBtb2RlbF9uYW1lPSd0aWNrZXRjaGFuZ2UnLAogICAgICAgICAgICBuYW1lPSdjb21wbGV0ZWRBdFRvJywKICAgICAgICAgICAgZmllbGQ9bW9kZWxzLkRhdGVUaW1lRmllbGQoYmxhbms9VHJ1ZSwgbnVsbD1UcnVlKSwKICAgICAgICApLAogICAgICAgIG1pZ3JhdGlvbnMuQWRkRmllbGQoCiAgICAgICAgICAgIG1vZGVsX25hbWU9J3RpY2tldGhpc3RvcnlyZWNvcmQnLAogICAgICAgICAgICBuYW1lPSdjb21wbGV0ZWRBdCcsCiAgICAgICAgICAgIGZpZWxkPW1vZGVscy5EYXRlVGltZUZpZWxkKGJsYW5rPVRydWUsIG51bGw9VHJ1ZSksCiAgICAgICAgKSwKICAgIF0K
+
+exports.up = async (knex) => {
+    await knex.raw(`
+    BEGIN;
+--
+-- Add field completedAt to ticket
+--
+ALTER TABLE "Ticket" ADD COLUMN "completedAt" timestamp with time zone NULL;
+--
+-- Add field completedAtFrom to ticketchange
+--
+ALTER TABLE "TicketChange" ADD COLUMN "completedAtFrom" timestamp with time zone NULL;
+--
+-- Add field completedAtTo to ticketchange
+--
+ALTER TABLE "TicketChange" ADD COLUMN "completedAtTo" timestamp with time zone NULL;
+--
+-- Add field completedAt to tickethistoryrecord
+--
+ALTER TABLE "TicketHistoryRecord" ADD COLUMN "completedAt" timestamp with time zone NULL;
+--
+-- Fill tickets completedAt field
+--
+UPDATE "Ticket"
+SET "completedAt" = "TicketCompletedTime"."completedAt"
+FROM (
+-- select last completed time when ticket was completed
+         SELECT t.id as "ticketId", MAX(tc."createdAt") as "completedAt"
+         FROM "Ticket" t
+                  JOIN "TicketChange" tc
+                       ON t.id = tc.ticket AND tc."statusDisplayNameTo" = 'ticket.status.COMPLETED.name'
```

Comment:
It's better to compare by `statusIdTo`, because translated status names was introduced after `TicketChange`
In production database we have status names without translations.

Following SQL-query against production database returns following results:
```
select tc."statusDisplayNameTo" from "TicketChange" as tc
group by "statusDisplayNameTo"
```

```
ticket.status.CLOSED.name
ticket.status.COMPLETED.name
ticket.status.DECLINED.name
ticket.status.DEFERRED.name
ticket.status.IN_PROGRESS.name
ticket.status.OPEN.name
В работе
Выполнена
Закрыта
Открыта
Отложена
Отменена
```

Reply:
used `statusIdTo` in join-condition